File: 66a.py
class UnionFinder:
    def __init__(self, n_node):
        self.nodes = [UnionFinder.Node(i) for i in range(n_node+1)]

    def __root(self, id):
        cur = self.nodes[id]
        while cur.parent != None:
            cur = cur.parent
        return cur

    def is_connected(self, id1, id2):
        return self.__root(id1).id == self.__root(id2).id

    def unite(self, id1, id2):
        r1 = self.__root(id1)
        r2 = self.__root(id2)

        if r1.id == r2.id:
            return

        if r1.nchild > r2.nchild:
            r1, r2 = r2, r1

        r2.nchild += r1.nchild
        r1.parent = r2

    class Node:
        def __init__(self, _id):
            self.id = _id
            self.parent = None
            self.nchild = 1


def solve_a66(n_node, queries):
    uf = UnionFinder(n_node)

    for q in queries:
        n1, n2 = q[1], q[2]

        if q[0] == 1:
            uf.unite(n1, n2)
        else:
            print("Yes" if uf.is_connected(n1, n2) else "No")


def main():
    N, Q = map(int, input().split())
    queries = []
    for _ in range(Q):
        q = list(map(int, input().split()))
        queries.append(q)

    solve_a66(N, queries)


main()

File: 66a.cc
#include <iostream>
#include <vector>

using u64 = uint64_t;

class UnionFinder {
  public:
    UnionFinder(u64 n_node) {
        std::vector<Node> _nodes;

        for (u64 i = 0; i <= n_node; i++)
            _nodes.push_back(Node(i));

        nodes = _nodes;
    }

    void unite(u64 id1, u64 id2) {
        Node *r1 = root(id1);
        Node *r2 = root(id2);

        r1->parent = r2;
    }

    bool is_connected(u64 id1, u64 id2) {
        Node *r1 = root(id1);
        Node *r2 = root(id2);

        return r1->id == r2->id;
    }

  private:
    class Node {
      public:
        Node(u64 id) : id(id), parent(nullptr) {}

        u64 id;
        Node *parent;
    };

    Node *root(u64 id) {
        Node *cur = &nodes[id];
        if (cur->parent == nullptr)
            return cur;

        cur->parent = root(cur->parent->id);
        return cur->parent;
    }

    std::vector<Node> nodes;
};

using query = std::vector<u64>;

void solve_a66(uint64_t N, std::vector<query> &queries) {
    UnionFinder uf = UnionFinder(N);

    for (u64 i = 0; i < queries.size(); i++) {
        u64 id1 = queries[i][1];
        u64 id2 = queries[i][2];

        if (queries[i][0] == 1) {
            uf.unite(id1, id2);
        } else {
            bool b = uf.is_connected(id1, id2);
            printf("%s\n", b ? "Yes" : "No");
        }
    }
}

int main() {
    u64 N, Q;
    std::vector<query> queries;

    std::cin >> N >> Q;
    for (u64 i = 0; i < Q; i++) {
        u64 q, n1, n2;
        std::cin >> q >> n1 >> n2;
        queries.push_back({q, n1, n2});
    }

    solve_a66(N, queries);
}